src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -696,24 +696,22 @@ private enum EnumeratorResult
             FailedAndReported
         }
 
-        private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder builder, BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics)
+        private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics)
         {
             TypeSymbol collectionExprType = collectionExpr.Type;
 
-            if (collectionExpr.ConstantValue != null)
+            if (collectionExprType is null) // There's no way to enumerate something without a type.
             {
-                if (collectionExpr.ConstantValue.IsNull)
+                if (collectionExpr.ConstantValue is { IsNull: true })
```

Comment:
Why did this first section need to be refactored? Consider reverting.

Reply:
Ah, I see. Nevermind.

---
In reply to: [443811375](https://github.com/dotnet/roslyn/pull/45160#discussion_r443811375) [](ancestors = 443811375)